components/brave_wallet/browser/solana_message.cc

```diff
@@ -104,91 +252,112 @@ void SolanaMessage::GetUniqueAccountMetas(
-  for (const auto& account_meta : unique_account_metas) {
-    if (account_meta.is_signer) {
-      if (signers)
-        signers->push_back(account_meta.pubkey);
-      if (num_required_signatures == UINT8_MAX)
-        return absl::nullopt;
-      num_required_signatures++;
-      if (!account_meta.is_writable) {
-        if (num_readonly_signed_accounts == UINT8_MAX)
-          return absl::nullopt;
-        num_readonly_signed_accounts++;
-      }
-    } else if (!account_meta.is_writable) {
-      if (num_readonly_unsigned_accounts == UINT8_MAX)
-        return absl::nullopt;
-      num_readonly_unsigned_accounts++;
-    }
-  }
-  message_bytes.push_back(num_required_signatures);
-  message_bytes.push_back(num_readonly_signed_accounts);
-  message_bytes.push_back(num_readonly_unsigned_accounts);
+  message_bytes.emplace_back(message_header_.num_required_signatures);
+  message_bytes.emplace_back(message_header_.num_readonly_signed_accounts);
+  message_bytes.emplace_back(message_header_.num_readonly_unsigned_accounts);
 
   // Compact array of account addresses.
-  CompactU16Encode(unique_account_metas.size(), &message_bytes);
-  for (const auto& account_meta : unique_account_metas) {
-    std::vector<uint8_t> pubkey(kSolanaPubkeySize);
-    if (!Base58Decode(account_meta.pubkey, &pubkey, pubkey.size()))
-      return absl::nullopt;
-    message_bytes.insert(message_bytes.end(), pubkey.begin(), pubkey.end());
+  CompactU16Encode(static_account_keys_.size(), &message_bytes);
+  for (size_t i = 0; i < static_account_keys_.size(); ++i) {
+    message_bytes.insert(message_bytes.end(),
+                         static_account_keys_[i].bytes().begin(),
+                         static_account_keys_[i].bytes().end());
+
+    if (signers && i < message_header_.num_required_signatures) {
+      signers->emplace_back(static_account_keys_[i].ToBase58());
+    }
   }
 
   // Recent blockhash.
   std::vector<uint8_t> recent_blockhash_bytes(kSolanaBlockhashSize);
   if (!Base58Decode(recent_blockhash_, &recent_blockhash_bytes,
-                    recent_blockhash_bytes.size()))
+                    recent_blockhash_bytes.size())) {
     return absl::nullopt;
+  }
   message_bytes.insert(message_bytes.end(), recent_blockhash_bytes.begin(),
                        recent_blockhash_bytes.end());
 
   // Compact array of instructions.
   CompactU16Encode(instructions_.size(), &message_bytes);
   for (const auto& instruction : instructions_) {
-    if (!instruction.Serialize(unique_account_metas, &message_bytes))
+    auto compiled_instruction = SolanaCompiledInstruction::FromInstruction(
+        instruction, static_account_keys_, address_table_lookups_,
+        num_of_write_indexes, num_of_read_indexes);
+    if (!compiled_instruction) {
       return absl::nullopt;
+    }
+    compiled_instruction->Serialize(&message_bytes);
+  }
+
+  // Compact array of address table lookups.
+  if (version_ == mojom::SolanaMessageVersion::kV0) {
```

Comment:
No we are not, I personally prefer letting Solana backend to define what transactions are invalid and rejected the requests. I would prefer not to duplicate some checks on our side, as we might be out-of-sync with Solana backend.

Reply:
Ok, I'm good with that approach for now to prevent unnecessary interop issues. We may find in the future that we want to perform this validation (e.g. reduce mempool spam/invalid RPC requests) then we can handle it specifically for the cases we want to address.